### PR TITLE
Cryptography.Xml.Tests: skip tests that require rsa+sha1 when platform doesn't support it.

### PR DESCRIPTION
**Pull request by tmds**
This causes 37 tests that fail on RHEL9 due to not supporting rsa+sha1 to be skipped.

@bartonjs @vcsjones ptal.

### CODE REVIEW

File: src/libraries/System.Security.Cryptography.Xml/tests/SignatureSupport.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        private static int _supportsRsaSha1Signatures = 0;
+
+        public static bool SupportsRsaSha1Signatures
+        {
+            get
+            {
+                if (_supportsRsaSha1Signatures == 0)
+                {
+                    bool supported = System.Security.Cryptography.Tests.SignatureSupport.CanProduceSha1Signature(new RSACryptoServiceProvider());
```
Comment (danmoseley) [inline src/libraries/System.Security.Cryptography.Xml/tests/SignatureSupport.cs:13]: Is it possible to imagine a product bug that broke the functionality would cause this to fail, causing the tests you annotated to not run on any platform, concealing the bug?

File: src/libraries/System.Security.Cryptography.Xml/tests/SignatureSupport.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+namespace System.Security.Cryptography.Xml.Tests
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography.Xml/tests/SignatureSupport.cs:1]: License header is missing.

```suggestion
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

namespace System.Security.Cryptography.Xml.Tests
```

File: src/libraries/System.Security.Cryptography.Xml/tests/SignatureSupport.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+                if (_supportsRsaSha1Signatures == 0)
+                {
+                    bool supported = System.Security.Cryptography.Tests.SignatureSupport.CanProduceSha1Signature(new RSACryptoServiceProvider());
+                    _supportsRsaSha1Signatures = supported ? 1 : -1;
+                }
+                return _supportsRsaSha1Signatures == 1;
+            }
+        }   
+    }
+}
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography.Xml/tests/SignatureSupport.cs:20]: Nit: Blank lines at the end of the file are recommended.

https://github.com/dotnet/runtime/blob/fff4ff0913a2396cfae61c701f2b59ffe5e27b49/.editorconfig#L9-L10

```suggestion
}

```

File: src/libraries/System.Security.Cryptography.Xml/tests/SignatureSupport.cs (hunk 2 above)
Comment (vcsjones) [reply src/libraries/System.Security.Cryptography.Xml/tests/SignatureSupport.cs:1]: (I thought we had a lint for that)

File: src/libraries/System.Security.Cryptography.Xml/tests/SignatureSupport.cs (hunk 1 above)
Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.Xml/tests/SignatureSupport.cs:16]: This library should just include that same file from Common.

File: src/libraries/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs [hunk 4]
```diff
@@ -441,7 +441,7 @@ public void GetIdElement()
             Assert.True(xel.InnerXml.StartsWith("<ObjectListTag"), "GetIdElement");
         }
 
-        [Fact]
+        [ConditionalFact(typeof(SignatureSupport), nameof(SignatureSupport.SupportsRsaSha1Signatures))]
         public void GetPublicKey()
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs:445]: What fails about this test?

That's another aspect of the general "these tests should be indicating what the behavior is with this OS configuration, instead of being turned off"

**Review by bartonjs — CHANGES_REQUESTED** [review]
The tests shouldn't be suppressed, where at all possible.  Instead, the tests should be indicating the expected behavior.

File: src/libraries/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs (hunk 4 above)
Comment (tmds) [reply src/libraries/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs:445]: Nothing. I had wrongly assumed any test that uses `signature` (which has `http://www.w3.org/2000/09/xmldsig#rsa-sha1`) would fail.

**Comment by tmds** [comment]
These tests fail by throwing `Interop+Crypto+OpenSslCryptographicException : error:03000098:digital envelope routines::invalid digest`.

```
at Interop.Crypto.RsaVerifyHash(SafeEvpPKeyHandle pkey, RSASignaturePaddingMode paddingMode, IntPtr digestAlgorithm, ReadOnlySpan`1 hash, ReadOnlySpan`1 signature) in /home/tester/runtime/src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EvpPkey.Rsa.cs:line 190 
at System.Security.Cryptography.RSAOpenSsl.VerifyHash(ReadOnlySpan`1 hash, ReadOnlySpan`1 signature, HashAlgorithmName hashAlgorithm, RSASignaturePadding padding) in /home/tester/runtime/src/libraries/Common/src/System/Security/Cryptography/RSAOpenSsl.cs:line 833 
at System.Security.Cryptography.RSAOpenSsl.VerifyHash(Byte[] hash, Byte[] signature, HashAlgorithmName hashAlgorithm, RSASignaturePadding padding) in /home/tester/runtime/src/libraries/Common/src/System/Security/Cryptography/RSAOpenSsl.cs:line 822 
at System.Security.Cryptography.RSAWrapper.VerifyHash(Byte[] hash, Byte[] signature, HashAlgorithmName hashAlgorithm, RSASignaturePadding padding) in /home/tester/runtime/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RSAWrapper.cs:line 66 
at System.Security.Cryptography.RSAPKCS1SignatureDeformatter.VerifySignature(Byte[] rgbHash, Byte[] rgbSignature) in /home/tester/runtime/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RSAPKCS1SignatureDeformatter.cs:line 55 
at System.Security.Cryptography.Xml.SignedXml.CheckSignedInfo(AsymmetricAlgorithm key) in /home/tester/runtime/src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SignedXml.cs:line 1035 
at System.Security.Cryptography.Xml.SignedXml.CheckSignature(AsymmetricAlgorithm key) in /home/tester/runtime/src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SignedXml.cs:line 290 
at System.Security.Cryptography.Xml.SignedXml.CheckSignatureReturningKey(AsymmetricAlgorithm& signingKey) in /home/tester/runtime/src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SignedXml.cs:line 274 
at System.Security.Cryptography.Xml.SignedXml.CheckSignature() in /home/tester/runtime/src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SignedXml.cs:line 253 
at System.Security.Cryptography.Xml.Tests.SignedXmlTest.CheckSignatureHandlesIncorrectOrTamperedReferenceWithMultipleEnvelopedSignatures(String signatureParent, String newReference) in /home/tester/runtime/src/libraries/System.Security.Cryptography.Xml/tests/SignedXmlTest.cs:line 1858 
at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor) 
at System.Reflection.RuntimeMethodInfo.InvokeNonEmitUnsafe(Object obj, IntPtr* arguments, Span`1 argsForTemporaryMonoSupport, BindingFlags invokeAttr) in /home/tester/runtime/src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.CoreCLR.cs:line 383
```

> The tests shouldn't be suppressed, where at all possible. Instead, the tests should be indicating the expected behavior.

@bartonjs rather than being skipped, you prefer these tests to be updated so they verify `CryptographicException` is thrown when `!SignatureSupport.SupportsRsaSha1Signatures`?

File: src/libraries/Common/tests/System/Security/Cryptography/SignatureSupport.cs [hunk 5]
```diff
@@ -1,51 +1,69 @@
...
+            get
+            {
+                if (_supportsRsaSha1Signatures == 0)
+                {
+                    bool supported = CanProduceSha1Signature(new RSACryptoServiceProvider());
+                    _supportsRsaSha1Signatures = supported ? 1 : -1;
+                }
+                return _supportsRsaSha1Signatures == 1;
+            }
+        }
```
Comment (vcsjones) [inline src/libraries/Common/tests/System/Security/Cryptography/SignatureSupport.cs:67]: My preference would be for this to be as localized as possible to S.S.C.Xml. An implementation should normally be passed in because their behavior might be different, and we support multiple providers on a single platform. For example, `RSAOpenSsl` and `RSA.Create()` on macOS use two entirely different RSA implementations.

So tests should be written with that in mind. Since S.S.C.Xml usually uses `RSACryptoServiceProvider` internally to do its RSA-y stuff, this makes sense there. But I wouldn't want this to be a defact-o way to determine "Is RSA supported or not".

tl;dr: I think this should be in `S.S.C.Xml.Tests` and renamed to `SupportsRsaSha1CspSignatures` or something.

Comment (tmds) [reply src/libraries/Common/tests/System/Security/Cryptography/SignatureSupport.cs:67]: I think @bartonjs suggested to use the same file for both: https://github.com/dotnet/runtime/pull/69713#discussion_r880826222.

Comment (bartonjs) [reply src/libraries/Common/tests/System/Security/Cryptography/SignatureSupport.cs:67]: I neglected to pay attention to the fact that the shared file requires an implementation to be passed in when I made my comment.

I agree with Kevin's comments here and withdraw my prior one.

**Comment by bartonjs** [comment]
I like the changes to the tests; but apparently the SignatureSupport changes should be popped to look like they did before I made a misinformed comment.

File: src/libraries/System.Security.Cryptography.Xml/tests/SignatureSupport.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+                        supported = true;
+                    }
+                    catch (CryptographicException)
+                    {
+                        supported = false;
+                    }
+
+                    _supportsRsaSha1Signatures = supported ? 1 : -1;
+                }
+                return _supportsRsaSha1Signatures == 1;
```
Comment (tmds) [inline src/libraries/System.Security.Cryptography.Xml/tests/SignatureSupport.cs:33]: @vcsjones @bartonjs does this look good?

**Comment by tmds** [comment]
I verified that `Cryptography.Xml.Tests` passes with these changes on RHEL9.

**Review by bartonjs — APPROVED** [review]
